commit 0dd8486b5cfe8048e0613334659d9252ecd1b08a
Author: Linus Torvalds <torvalds@linux-foundation.org>
Date:   Sat Aug 1 11:18:56 2009 -0700

    do_sigaltstack: small cleanups
    
    The previous commit ("do_sigaltstack: avoid copying 'stack_t' as a
    structure to user space") fixed a real bug.  This one just cleans up the
    copy from user space to that gcc can generate better code for it (and so
    that it looks the same as the later copy back to user space).
    
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Adjusted to apply to Debian's 2.6.30 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.30.orig/kernel/signal.c linux-source-2.6.30/kernel/signal.c
--- linux-source-2.6.30.orig/kernel/signal.c	2009-08-14 18:04:08.000000000 -0600
+++ linux-source-2.6.30/kernel/signal.c	2009-08-14 18:05:13.000000000 -0600
@@ -2424,10 +2424,12 @@ do_sigaltstack (const stack_t __user *us
 		int ss_flags;
 
 		error = -EFAULT;
-		if (!access_ok(VERIFY_READ, uss, sizeof(*uss))
-		    || __get_user(ss_sp, &uss->ss_sp)
-		    || __get_user(ss_flags, &uss->ss_flags)
-		    || __get_user(ss_size, &uss->ss_size))
+		if (!access_ok(VERIFY_READ, uss, sizeof(*uss)))
+			goto out;
+		error = __get_user(ss_sp, &uss->ss_sp) |
+			__get_user(ss_flags, &uss->ss_flags) |
+			__get_user(ss_size, &uss->ss_size);
+		if (error)
 			goto out;
 
 		error = -EPERM;
